// apollo/lib/commands/service/download.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const graphql_1 = require("graphql");
const fs_1 = require("fs");
const Command_1 = require("../../Command");
class ServiceDownload extends Command_1.ProjectCommand {
    async run() {
        let result;
        let gitContext;
        await this.runTasks(({ args, project, flags }) => [
            {
                title: `Saving schema to ${args.output}`,
                task: async () => {
                    const schema = await project.resolveSchema({ tag: flags.tag });
                    fs_1.writeFileSync(args.output, JSON.stringify(graphql_1.introspectionFromSchema(schema), null, 2));
                }
            }
        ]);
    }
}
ServiceDownload.aliases = ["schema:download"];
ServiceDownload.description = "Download the schema from your GraphQL endpoint.";
ServiceDownload.flags = Object.assign({}, Command_1.ProjectCommand.flags, { tag: command_1.flags.string({
        char: "t",
        description: "The published tag to check this service against",
        default: "current"
    }), skipSSLValidation: command_1.flags.boolean({
        char: "k",
        description: "Allow connections to an SSL site without certs"
    }) });
ServiceDownload.args = [
    {
        name: "output",
        description: "Path to write the introspection result to",
        required: true,
        default: "schema.json"
    }
];
exports.default = ServiceDownload;
